ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java

```diff
@@ -575,6 +581,154 @@ private JoinNode buildJoin(final Join root, final String prefix, final boolean i
     );
   }
 
+  private void verifyJoin(final JoinInfo joinInfo,
+                          final PlanNode leftNode,
+                          final PlanNode rightNode) {
+    final JoinType joinType = joinInfo.getType();
+    final Expression leftExpression = joinInfo.getLeftJoinExpression();
+    final Expression rightExpression = joinInfo.getRightJoinExpression();
+
+    if (leftNode.getNodeOutputType() == DataSourceType.KSTREAM) {
+      if (rightNode.getNodeOutputType() == DataSourceType.KTABLE) {
+
+        // stream-table join detected
+
+        if (joinType.equals(JoinType.OUTER)) {
+          throw new KsqlException(String.format(
+                  "Invalid join type: "
+                      + "full-outer join not supported for stream-table join: %s %s %s.",
+                  joinInfo.getLeftSource().getDataSource().getName().text(),
+                  joinType,
+                  joinInfo.getRightSource().getDataSource().getName().text()
+              ));
+        }
+
+        if (joinOnNonKeyAttribute(rightExpression, rightNode)) {
+          throw new KsqlException(String.format(
+                  "Invalid join condition:"
+                      + " stream-table joins require to join on the table's primary key: %s = %s.",
+                  leftExpression,
+                  rightExpression
+              ));
+        }
+      }
+
+      // stream-stream join detected: nothing to verify
+
+    } else {
+      if (rightNode.getNodeOutputType() == DataSourceType.KSTREAM) {
+        throw new KsqlException(String.format(
+                "Invalid join order:"
+                  + " table-stream joins are not supported; only stream-table joins: %s %s %s.",
+                joinInfo.getLeftSource().getDataSource().getName().text(),
+                joinType,
+                joinInfo.getRightSource().getDataSource().getName().text()
+            ));
+      }
+
+      // table-table join detected
+
+      if (joinOnNonKeyAttribute(rightExpression, rightNode)) {
+        throw new KsqlException(String.format(
+                "Invalid join condition:"
+                    + " table-table joins require to join on the primary key of the right input"
+                    + " table: %s = %s.",
```

Comment:
This error message seems confusing. It's printing what the user passed in after the colon, so a user might think that they're already passing in a valid join condition even though that's not the case. Ideally we'd clarify by printing what the "primary key of the right input" is but that's more work. A minor clarification in the meantime could be the following:
```suggestion
                    + " table. Got: %s = %s.",
```

Reply:
> "primary key of the right input

I tried that but it's not so simple, because there might be "tree"...